### PR TITLE
contrib: refactor http request span tags

### PR DESCRIPTION
**Pull request by Julio-Guerra**
Refactor the start and finish of http request spans into functions
shared across contribs so that the expected set of http request span
tags are properly implemented.

The new standard `http.useragent` is added in this PR as a first example of span tag addition spread out to every HTTP integration.

### CODE REVIEW

File: contrib/net/http/trace.go [hunk 1]
```diff
@@ -41,20 +42,33 @@ type ServeConfig struct {
...
+
+	if appsec.Enabled() {
+		h = httpsec.WrapHandler(h, span, cfg.RouteParams)
 	}
+	h.ServeHTTP(rw, r.WithContext(ctx))
+}
+
+// StartRequestSpan starts an HTTP request span with the standard list of HTTP request span tags. URL query parameters
+// are added to the URL tag when queryParams is true. Any further span start option can be added with opts.
+func StartRequestSpan(r *http.Request, service, resource string, queryParams bool, opts ...ddtrace.StartSpanOption) (tracer.Span, context.Context) {
```
Comment (gbbr) [inline contrib/net/http/trace.go:59]: How will users benefit of this new API? Perhaps these should be internal.

Comment (Julio-Guerra) [reply contrib/net/http/trace.go:59]: Indeed, it's only intended for the contribs so it could be moved into a `contrib/internal` package.

File: contrib/net/http/trace.go [hunk 2]
```diff
@@ -41,20 +42,33 @@ type ServeConfig struct {
...
 	path := r.URL.Path
-	if cfg.QueryParams {
+	if queryParams {
 		path += "?" + r.URL.RawQuery
 	}
-	opts := append([]ddtrace.StartSpanOption{
+	opts = append([]ddtrace.StartSpanOption{
```
Comment (Julio-Guerra) [inline contrib/net/http/trace.go:64]: Should we always add `tracer.Measured()` as it is almost always present in the contribs?

File: contrib/net/http/trace.go [hunk 3]
```diff
@@ -41,20 +42,33 @@ type ServeConfig struct {
...
+		FinishRequestSpan(span, ddrw.status, cfg.FinishOpts...)
+	}()
+
+	if appsec.Enabled() {
+		h = httpsec.WrapHandler(h, span, cfg.RouteParams)
 	}
+	h.ServeHTTP(rw, r.WithContext(ctx))
+}
+
+// StartRequestSpan starts an HTTP request span with the standard list of HTTP request span tags. URL query parameters
```
Comment (katiehockman) [inline contrib/net/http/trace.go:57]: What is the "standard list of HTTP request span tags". Is that documented somewhere? If so, link to that here. If not, I suggest writing them out explicitly or providing more explanation in the godoc of this function.

File: contrib/net/http/trace.go (hunk 2 above)
Comment (katiehockman) [reply contrib/net/http/trace.go:64]: Same question for `FinishRequestSpan`. That sounds like something that shouldn't be part of the API.

File: contrib/net/http/trace.go [hunk 4]
```diff
@@ -63,24 +77,23 @@ func TraceAndServe(h http.Handler, w http.ResponseWriter, r *http.Request, cfg *
...
-			span.SetTag(ext.Error, fmt.Errorf("%d: %s", ddrw.status, http.StatusText(ddrw.status)))
-		}
-		span.Finish(cfg.FinishOpts...)
-	}()
+	return tracer.StartSpanFromContext(r.Context(), "http.request", opts...)
+}
 
-	if appsec.Enabled() {
-		h = httpsec.WrapHandler(h, span, cfg.RouteParams)
+// FinishRequestSpan finishes the given HTTP request span with the standard list of HTTP request span tags.
```
Comment (katiehockman) [inline contrib/net/http/trace.go:83]: Let's be clearer about what "finishes" means here. What exactly is it doing?

File: contrib/net/http/trace.go [hunk 5]
```diff
@@ -63,24 +77,23 @@ func TraceAndServe(h http.Handler, w http.ResponseWriter, r *http.Request, cfg *
...
+// Any further span finish option can be added with opts.
+func FinishRequestSpan(s tracer.Span, status int, opts ...tracer.FinishOption) {
+	var statusStr string
+	if status == 0 {
+		statusStr = "200"
+	} else {
+		statusStr = strconv.Itoa(status)
 	}
-	h.ServeHTTP(rw, r.WithContext(ctx))
+	s.SetTag(ext.HTTPCode, statusStr)
```
Comment (katiehockman) [inline contrib/net/http/trace.go:92]: nit: I preferred the way the code was previously where `s.SetTag` is within the if statement. It was a bit clearer to read that way in my opinion.

File: contrib/net/http/trace.go [hunk 6]
```diff
@@ -63,24 +77,23 @@ func TraceAndServe(h http.Handler, w http.ResponseWriter, r *http.Request, cfg *
...
+// FinishRequestSpan finishes the given HTTP request span with the standard list of HTTP request span tags.
+// Any further span finish option can be added with opts.
+func FinishRequestSpan(s tracer.Span, status int, opts ...tracer.FinishOption) {
+	var statusStr string
+	if status == 0 {
+		statusStr = "200"
+	} else {
+		statusStr = strconv.Itoa(status)
 	}
-	h.ServeHTTP(rw, r.WithContext(ctx))
```
Comment (katiehockman) [inline contrib/net/http/trace.go:83]: Where did these lines go? Should they be in `FinishRequestSpan`?

File: contrib/net/http/trace.go [hunk 7]
```diff
@@ -41,20 +42,33 @@ type ServeConfig struct {
...
-		cfg = new(ServeConfig)
+	span, ctx := StartRequestSpan(r, cfg.Service, cfg.Resource, cfg.QueryParams, cfg.SpanOpts...)
+	rw, ddrw := wrapResponseWriter(w)
+	defer func() {
+		FinishRequestSpan(span, ddrw.status, cfg.FinishOpts...)
+	}()
+
+	if appsec.Enabled() {
+		h = httpsec.WrapHandler(h, span, cfg.RouteParams)
 	}
```
Comment (katiehockman) [inline contrib/net/http/trace.go:53]: This seems like an important check to keep. Are we risking a nil-pointer exception now?
(If there isn't a test for a nil cfg being passed to this function, then let's add one)

**Review by katiehockman — COMMENTED** [review]
Overall looks like a helpful refactor 👍

File: contrib/net/http/trace.go (hunk 1 above)
Comment (Julio-Guerra) [reply contrib/net/http/trace.go]: I created `contrib/internal/httptrace` and moved those functions in there: https://github.com/DataDog/dd-trace-go/pull/1286/commits/41b9cbe0dae9a5d2f48e914be4972c56d11d5902

File: contrib/emicklei/go-restful/restful.go [hunk 8]
```diff
@@ -26,51 +26,29 @@ func FilterFunc(configOpts ...Option) restful.FilterFunction {
...
-		if spanctx, err := tracer.Extract(tracer.HTTPHeadersCarrier(req.Request.Header)); err == nil {
-			opts = append(opts, tracer.ChildOf(spanctx))
+			opts = []ddtrace.StartSpanOption{tracer.Tag(ext.EventSampleRate, cfg.analyticsRate)}
 		}
-		span, ctx := tracer.StartSpanFromContext(req.Request.Context(), "http.request", opts...)
-		defer span.Finish()
+		span, ctx := httptrace.StartRequestSpan(req.Request, cfg.serviceName, req.SelectedRoutePath(), false, opts...)
+		defer func() {
+			httptrace.FinishRequestSpan(span, resp.StatusCode(), tracer.WithError(resp.Error()))
+		}()
```
[the remark this answers is not in the document]
Comment (Julio-Guerra) [reply contrib/emicklei/go-restful/restful.go:36]: No, this allows to also defer the arguments of the call such as getting the status code after the user handler was called.

File: contrib/net/http/trace.go (hunk 3 above)
Comment (Julio-Guerra) [reply contrib/net/http/trace.go]: The document is internal in our confluence: https://datadoghq.atlassian.net/wiki/spaces/APM/pages/2357395856/Span+attributes

What I personally think would be cleaner and better would be to move all the HTTP tag names currently defined in `ddtrace/ext/tags.go` into a const block of the new internal httptrace package so that users of this package can clearly see their list and documentation:

```go
// List of standard HTTP request span tags
const (
	// Method of the HTTP request.
	Method = "http.method"
	// StatusCode is the client the response status code the HTTP client received or the HTTP server sent.
	StatusCode = "http.status_code"
	// ... etc ...
)
```

File: contrib/net/http/trace.go (hunk 7 above)
Comment (Julio-Guerra) [reply contrib/net/http/trace.go]: My bad, I probably removed it by accident.
Restored by https://github.com/DataDog/dd-trace-go/pull/1286/commits/c2d56a854d2b8e306da303a09b6d95dd79a42d44

File: contrib/net/http/trace.go (hunk 2 above)
Comment (Julio-Guerra) [reply contrib/net/http/trace.go]: Now moved into an internal package in https://github.com/DataDog/dd-trace-go/pull/1286/commits/029b8b5b8f134c893550392a960ffbb4a10ac264

File: contrib/internal/httptrace/httptrace.go [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+		path += "?" + r.URL.RawQuery
+	}
+	opts = append([]ddtrace.StartSpanOption{
+		tracer.SpanType(ext.SpanTypeWeb),
+		tracer.ServiceName(service),
+		tracer.ResourceName(resource),
+		tracer.Tag(ext.HTTPMethod, r.Method),
+		tracer.Tag(ext.HTTPURL, path),
+		tracer.Tag(ext.HTTPUserAgent, r.UserAgent()),
+	}, opts...)
```
Comment (Julio-Guerra) [inline contrib/internal/httptrace/httptrace.go:35]: Should we also add the option `tracer.Measured()` too here?
Almost every integration is setting it so I was wondering what would be the impact of having it by default?
It would remove a few more lines from the integrations by putting it in common here.

File: contrib/net/http/trace.go (hunk 6 above)
Comment (Julio-Guerra) [reply contrib/net/http/trace.go]: This line moved and I think you got confused with the unified git diff view.
But now that Start/FinishRequestSpan were moved into their own internal package, the diff is now much clearer to read.

I'm resolving this comment and let you walk through the new diff of net/http/trace.go and let me know if anything 🙏

File: contrib/net/http/trace.go (hunk 5 above)
Comment (Julio-Guerra) [reply contrib/net/http/trace.go]: I was about to restore it until I realized we were converting the status code into a string twice in that function.
So I reused `statusStr` in the second place in https://github.com/DataDog/dd-trace-go/pull/1286/commits/41f4c74ae4dfd7847f14144baf18819f8abeb198

WDYT?

File: contrib/net/http/trace.go (hunk 4 above)
Comment (Julio-Guerra) [reply contrib/net/http/trace.go]: Added a small clarification in https://github.com/DataDog/dd-trace-go/pull/1286/commits/8d9bd76eb8b97263ace3e18f76774f6a74bb16c3

Note that the documentation of [`tracer.(Span).Finish()`](https://pkg.go.dev/gopkg.in/DataDog/dd-trace-go.v1@v1.38.1/ddtrace#Span)  doesn't say give much more details but that's where I would put extra details (out of the scope of this PR though).

WDYT?

File: contrib/labstack/echo/echotrace.go [hunk 10]
```diff
@@ -27,30 +27,26 @@ func Middleware(opts ...Option) echo.MiddlewareFunc {
...
-			}
+
 			var finishOpts []tracer.FinishOption
 			if cfg.noDebugStack {
-				finishOpts = append(finishOpts, tracer.NoDebugStack())
+				finishOpts = []tracer.FinishOption{tracer.NoDebugStack()}
 			}
-			span, ctx := tracer.StartSpanFromContext(request.Context(), "http.request", opts...)
-			defer func() { span.Finish(finishOpts...) }()
+
```
Comment (gbbr) [inline contrib/labstack/echo/echotrace.go:45]: Same here, and everywhere else. https://github.com/DataDog/dd-trace-go/wiki/Style-guidelines#general-guidelines

File: ddtrace/ext/tags.go [hunk 11]
```diff
@@ -33,6 +33,9 @@ const (
 	// HTTPURL sets the HTTP URL for a span.
 	HTTPURL = "http.url"
 
+	// HTTPUserAgent sets the HTTP user agent tag.
```
Comment (gbbr) [inline ddtrace/ext/tags.go:36]:
```suggestion
	// HTTPUserAgent specifies the name of the tag which holds the HTTP user agent tag.
```
It doesn't _set it_ per se, it is not a function.

File: contrib/net/http/trace.go (hunk 3 above)
Comment (katiehockman) [reply contrib/net/http/trace.go]: SGTM 👍

File: contrib/net/http/trace.go (hunk 4 above)
Comment (katiehockman) [reply contrib/net/http/trace.go]: I actually think the previous version was a bit better. I was thinking more about an explanation of the status strings and errors in the tags, but no biggie either way.

File: contrib/net/http/trace.go (hunk 5 above)
Comment (katiehockman) [reply contrib/net/http/trace.go]: Ah yep that's fine then 👍

File: contrib/emicklei/go-restful/restful.go [hunk 12]
```diff
@@ -26,51 +26,29 @@ func FilterFunc(configOpts ...Option) restful.FilterFunction {
...
-	}
-	if spanctx, err := tracer.Extract(tracer.HTTPHeadersCarrier(req.Request.Header)); err == nil {
-		opts = append(opts, tracer.ChildOf(spanctx))
-	}
-	span, ctx := tracer.StartSpanFromContext(req.Request.Context(), "http.request", opts...)
-	defer span.Finish()
+	span, ctx := httptrace.StartRequestSpan(req.Request, req.SelectedRoutePath(), req.SelectedRoutePath(), false)
+	defer func() {
+		httptrace.FinishRequestSpan(span, resp.StatusCode(), tracer.WithError(resp.Error()))
+	}()
```
Comment (gbbr) [inline contrib/emicklei/go-restful/restful.go:49]:
```suggestion
	defer httptrace.FinishRequestSpan(span, resp.StatusCode(), tracer.WithError(resp.Error()))
```
Or is it possible that `chain.ProcessFilter` changes the return values of `resp.StatusCode()` and `resp.Error()`. I assume that shouldn't be the case because the response is already final isn't it?

File: contrib/gin-gonic/gin/gintrace.go [hunk 13]
```diff
@@ -34,27 +33,21 @@ func Middleware(service string, opts ...Option) gin.HandlerFunc {
...
-		span, ctx := tracer.StartSpanFromContext(c.Request.Context(), "http.request", opts...)
-		defer span.Finish()
+		span, ctx := httptrace.StartRequestSpan(c.Request, cfg.serviceName, cfg.resourceNamer(c), false, opts...)
 
 		// pass the span through the request context
 		c.Request = c.Request.WithContext(ctx)
 
+		defer func() {
+			httptrace.FinishRequestSpan(span, c.Writer.Status())
+		}()
```
Comment (gbbr) [inline contrib/gin-gonic/gin/gintrace.go:49]:
```suggestion
		defer httptrace.FinishRequestSpan(span, c.Writer.Status())
```
Same question here. Also, why not put the defer just below the `StartRequestSpan` ? Is there anything important that `c.Request.WithContext(ctx)` does which changes the output of `c.Writer().Status()`?

If possible, it's nice to group the `defer` together with the thing it relates to.

File: contrib/go-chi/chi.v5/chi.go [hunk 14]
```diff
@@ -36,22 +35,23 @@ func Middleware(opts ...Option) func(next http.Handler) http.Handler {
...
-			}
-			opts = append(opts, cfg.spanOpts...)
-			span, ctx := tracer.StartSpanFromContext(r.Context(), "http.request", opts...)
-			defer span.Finish()
+			span, ctx := httptrace.StartRequestSpan(r, cfg.serviceName, "", false, opts...)
+
+			r = r.WithContext(ctx)
+			ww := middleware.NewWrapResponseWriter(w, r.ProtoMajor)
+
+			defer func() {
```
Comment (gbbr) [inline contrib/go-chi/chi.v5/chi.go:47]:
```suggestion
			span, ctx := httptrace.StartRequestSpan(r, cfg.serviceName, "", false, opts...)
			r = r.WithContext(ctx)
			ww := middleware.NewWrapResponseWriter(w, r.ProtoMajor)
			defer func() {
```
I think that removing the empty lines really does make sense to make it obvious that this is a chunk which does one focused thing: start and defer the end of a span.

File: contrib/go-chi/chi/chi.go [hunk 15]
```diff
@@ -36,22 +35,24 @@ func Middleware(opts ...Option) func(next http.Handler) http.Handler {
...
 				opts = append(opts, tracer.Tag(ext.EventSampleRate, cfg.analyticsRate))
 			}
-			if spanctx, err := tracer.Extract(tracer.HTTPHeadersCarrier(r.Header)); err == nil {
-				opts = append(opts, tracer.ChildOf(spanctx))
-			}
 			opts = append(opts, cfg.spanOpts...)
-			span, ctx := tracer.StartSpanFromContext(r.Context(), "http.request", opts...)
-			defer span.Finish()
+			span, ctx := httptrace.StartRequestSpan(r, cfg.serviceName, "", false, opts...)
+
```
Comment (gbbr) [inline contrib/go-chi/chi/chi.go:44]: Same here, but this is just a nit, not feeling too strong about it.

File: contrib/internal/httptrace/httptrace.go [hunk 16]
```diff
@@ -0,0 +1,62 @@
...
+	"fmt"
+	"net/http"
+	"strconv"
+
+	"gopkg.in/DataDog/dd-trace-go.v1/ddtrace"
+	"gopkg.in/DataDog/dd-trace-go.v1/ddtrace/ext"
+	"gopkg.in/DataDog/dd-trace-go.v1/ddtrace/tracer"
+)
+
+// StartRequestSpan starts an HTTP request span with the standard list of HTTP request span tags. URL query parameters
```
Comment (gbbr) [inline contrib/internal/httptrace/httptrace.go:21]: As @katiehockman has also suggested, we should clarify what the "standard list" is or means, specifically. Yes, there is a risk of losing sync between the docs and the code, but perhaps if you make a bulleted list it will be harder to miss.

File: contrib/internal/httptrace/httptrace.go (hunk 9 above)
Comment (gbbr) [reply contrib/internal/httptrace/httptrace.go:35]: I don't see why not! If it's ever a problem, we can always refactor because this is internal code!

File: contrib/internal/httptrace/httptrace.go [hunk 17]
```diff
@@ -0,0 +1,62 @@
...
+// are added to the URL tag when queryParams is true. Any further span start option can be added with opts.
+func StartRequestSpan(r *http.Request, service, resource string, queryParams bool, opts ...ddtrace.StartSpanOption) (tracer.Span, context.Context) {
```
Comment (gbbr) [inline contrib/internal/httptrace/httptrace.go:23]: I have some ideas on simplifying this function signature and as a result the API.

### service & resource

While these two are mandatory, there is nothing in the function stopping us from giving an empty string. For that reason, I think it'd be much nicer if we removed both service & resource and instead just use `opts` because they already allow setting both of these. Additionally, perhaps it makes sense to log a message (at the "warn" or "debug" level) if service and/or resource are empty.

### queryParams

I would like to propose removing `queryParams` from here for multiple reasons:
* I think that bool args make for an ugly API even if that API is internal.
* There are way too many args on this function, contributing to the ugliness of its API (not your fault, it came like this, but let's improve it!)
* It is `false` 99% of the time, except for `net/http`

I'd propose to just remove it, and inside the `net/http` package, when the query params are needed, simply use `tracer.Tag(ext.HTTPURL, r.URL.Path + "?" + r.URL.RawQuery)`.

WDYT?

File: contrib/internal/httptrace/httptrace.go [hunk 18]
```diff
@@ -0,0 +1,62 @@
...
+	if spanctx, err := tracer.Extract(tracer.HTTPHeadersCarrier(r.Header)); err == nil {
+		opts = append(opts, tracer.ChildOf(spanctx))
+	}
+	return tracer.StartSpanFromContext(r.Context(), "http.request", opts...)
+}
+
+// FinishRequestSpan finishes the given HTTP request span with its Finish() method along with the standard list of HTTP
+// request span tags.
+// Any further span finish option can be added with opts.
+func FinishRequestSpan(s tracer.Span, status int, opts ...tracer.FinishOption) {
```
Comment (gbbr) [inline contrib/internal/httptrace/httptrace.go:50]: I have similar thoughts here regarding `status`, but not sure if it actually helps here or just makes usage more difficult.

**Review by gbbr — COMMENTED** [review]
A few more nits & thoughts.

File: contrib/emicklei/go-restful/restful.go (hunk 12 above)
Comment (Julio-Guerra) [reply contrib/emicklei/go-restful/restful.go:49]: The arguments would be immediately taken when the deferred call is being stacked and so you wouldn't get the status code nor the error after the call to the handler but rather before (so 0 and nil).

File: contrib/gin-gonic/gin/gintrace.go (hunk 13 above)
Comment (Julio-Guerra) [reply contrib/gin-gonic/gin/gintrace.go:49]: The only reason why I've moved this right after the span creation was to avoid the common mistake of using the request context right after like if it was replaced by the new one while it wasn't yet.

This code doesn't do that so I'm good to move it back below defer everywhere I've done that. Let me know.

File: contrib/emicklei/go-restful/restful.go (hunk 12 above)
Comment (Julio-Guerra) [reply contrib/emicklei/go-restful/restful.go:49]: Restored in https://github.com/DataDog/dd-trace-go/pull/1286/commits/7dff5a4b6ea21526f1f9cfa535614fa104de3c67

File: contrib/internal/httptrace/httptrace.go (hunk 9 above)
Comment (Julio-Guerra) [reply contrib/internal/httptrace/httptrace.go]: Done in https://github.com/DataDog/dd-trace-go/pull/1286/commits/fcd3cd2b64b61e8d8479db9ec4cecf60e6c76ec7

File: contrib/internal/httptrace/httptrace.go (hunk 17 above)
Comment (Julio-Guerra) [reply contrib/internal/httptrace/httptrace.go:23]: - service & resource: I don't see any tracer option to set the resource name. As for the service name, there's a side effect in the option setting the global service name.  Do you still think I should use it? What about the resource name?

- Query params: we can especially remove it now that we know we can have them by default.

File: contrib/internal/httptrace/httptrace.go (hunk 16 above)
Comment (Julio-Guerra) [reply contrib/internal/httptrace/httptrace.go:21]: On it yes 👍
We actually can't move the HTTP tags out of `ddtrace/ext/tags.go` since it would be a breaking API change. So I'll just add their list in there for now. But the list might be very long once you support all the tags so I still think some `const ()` block documentation would be better.

File: contrib/emicklei/go-restful/restful.go (hunk 12 above)
Comment (gbbr) [reply contrib/emicklei/go-restful/restful.go:49]: Do you mean that `chain.ProcessFilter` might change the response? I just noticed Katie asked the same question...

File: contrib/internal/httptrace/httptrace.go (hunk 17 above)
Comment (gbbr) [reply contrib/internal/httptrace/httptrace.go:23]: Oh, they are there: https://github.com/DataDog/dd-trace-go/blob/v1/ddtrace/tracer/option.go#L746-L755. They are actually used in this very function 🙂

File: contrib/internal/httptrace/httptrace.go (hunk 18 above)
Comment (gbbr) [reply contrib/internal/httptrace/httptrace.go:51]: Fine by me.

File: contrib/net/http/trace.go (hunk 3 above)
Comment (Julio-Guerra) [reply contrib/net/http/trace.go]: Done in https://github.com/DataDog/dd-trace-go/pull/1286/commits/e989bc6a2b14146ed4a2186b599b1aed413c7699
But note that I couldn't move the existing public definitions of the `ext` package to avoid breaking changes and so I redefinted them in httptrace for the sake of creating this documentation group.

File: contrib/net/http/trace.go (hunk 4 above)
Comment (Julio-Guerra) [reply contrib/net/http/trace.go]: Reverted in https://github.com/DataDog/dd-trace-go/pull/1286/commits/92607eff5eb294c9c25aabb2540543797efd8592

File: ddtrace/ext/tags.go (hunk 11 above)
Comment (Julio-Guerra) [reply ddtrace/ext/tags.go]: Fixed in https://github.com/DataDog/dd-trace-go/pull/1286/commits/e989bc6a2b14146ed4a2186b599b1aed413c7699

File: contrib/internal/httptrace/httptrace.go (hunk 16 above)
Comment (Julio-Guerra) [reply contrib/internal/httptrace/httptrace.go:34]: Done in https://github.com/DataDog/dd-trace-go/pull/1286/commits/e989bc6a2b14146ed4a2186b599b1aed413c7699

File: contrib/emicklei/go-restful/restful.go (hunk 12 above)
Comment (Julio-Guerra) [reply contrib/emicklei/go-restful/restful.go:49]: Yes, that's where the next handler is being called AFAIK, resulting in writing the response status code at some point. So the deferred block allows to get the response status code and error after the next filter has been called, when returning from the function.

File: contrib/internal/httptrace/httptrace.go (hunk 17 above)
Comment (Julio-Guerra) [reply contrib/internal/httptrace/httptrace.go]: Done in https://github.com/DataDog/dd-trace-go/pull/1286/commits/95a7b98715210af126f661f33e3d65ca25a3e66e

Note that we could go further by reworking the signature of `TraceAndServe` but the change got too big for the original purpose of this pull request which is about adding the new user agent tag to every HTTP integration.

File: contrib/go-chi/chi.v5/chi.go (hunk 14 above)
Comment (Julio-Guerra) [reply contrib/go-chi/chi.v5/chi.go]: Now fixed

File: contrib/labstack/echo/echotrace.go [hunk 19]
```diff
@@ -27,30 +27,26 @@ func Middleware(opts ...Option) echo.MiddlewareFunc {
 			fn(cfg)
 		}
 		log.Debug("contrib/labstack/echo: Configuring Middleware: %#v", cfg)
+
```
[the remark this answers is not in the document]
Comment (Julio-Guerra) [reply contrib/labstack/echo/echotrace.go]: Fixed

File: contrib/internal/httptrace/httptrace.go [hunk 20]
```diff
@@ -0,0 +1,80 @@
...
+	HTTPURL = ext.HTTPURL
+	// HTTPUserAgent is the user agent header value of the HTTP request.
+	HTTPUserAgent = "http.useragent"
+	// HTTPCode is the HTTP response status code sent by the HTTP request handler.
+	HTTPCode = ext.HTTPCode
+)
+
+// StartRequestSpan starts an HTTP request span with the standard list of HTTP request span tags. URL query parameters
+// are added to the URL tag when queryParams is true. Any further span start option can be added with opts.
+func StartRequestSpan(r *http.Request, queryParams bool, opts ...ddtrace.StartSpanOption) (tracer.Span, context.Context) {
```
Comment (gbbr) [inline contrib/internal/httptrace/httptrace.go:36]: Internal or not, I think `bool` args are not a good programming practice. There is only one single place in all of the usage in this repository where the `queryParams` param has any other value than `false`. I don't think that this one place warrants introducing a new parameter and an additional helper function. It isn't really de-duplicating anything.

Can we please get rid of the `queryParams` parameter and the `makeURLTag` helper? Instead, simply add the `tracer.Tag(HTTPURL, ...)` to `opts` in `net/http` when needed.

File: contrib/internal/httptrace/httptrace.go [hunk 21]
```diff
@@ -0,0 +1,80 @@
...
+	"gopkg.in/DataDog/dd-trace-go.v1/ddtrace/ext"
+	"gopkg.in/DataDog/dd-trace-go.v1/ddtrace/tracer"
+)
+
+// List of standard HTTP request span tags.
+const (
+	// HTTPMethod is the HTTP request method.
+	HTTPMethod = ext.HTTPMethod
+	// HTTPURL is the full HTTP request URL in the form `scheme://host[:port]/path[?query][#fragment]`.
+	HTTPURL = ext.HTTPURL
```
Comment (gbbr) [inline contrib/internal/httptrace/httptrace.go:27]: Why? Let's not. Aliases are used for refactoring and moving packages, not for convenience.

File: contrib/internal/httptrace/httptrace.go [hunk 22]
```diff
@@ -0,0 +1,80 @@
...
+	// HTTPUserAgent is the user agent header value of the HTTP request.
+	HTTPUserAgent = "http.useragent"
```
Comment (gbbr) [inline contrib/internal/httptrace/httptrace.go:29]: Let's add this to `ext` if it makes sense.

**Review by gbbr — CHANGES_REQUESTED** [review]

File: contrib/internal/httptrace/httptrace.go (hunk 20 above)
Comment (gbbr) [reply contrib/internal/httptrace/httptrace.go:36]: Actually it seems like we've already been through this https://github.com/DataDog/dd-trace-go/pull/1286#discussion_r873836690 🙂

Comment (Julio-Guerra) [reply contrib/internal/httptrace/httptrace.go:36]: > There is only one single place in all of the usage in this repository where the queryParams param has any other value than false.

If we follow the specification as specified today, we could remove this argument because the query params are supposed to be traced by default now, with an opt-out env var that we could check inside `httptrace`.

I wouldn't move the `http.url` tag creation out of `httptrace` to avoid the former copy/paste issue, which is the only purpose of this PR.

File: contrib/internal/httptrace/httptrace.go [hunk 23]
```diff
@@ -0,0 +1,80 @@
...
+	if status >= 500 && status < 600 {
+		s.SetTag(ext.Error, fmt.Errorf("%s: %s", statusStr, http.StatusText(status)))
+	}
+	s.Finish(opts...)
+}
+
+// Create the http.url value out of the given HTTP request.
+func makeURLTag(r *http.Request, queryParams bool) string {
+	var u strings.Builder
+	u.WriteString(r.URL.EscapedPath())
```
Comment (gbbr) [inline contrib/internal/httptrace/httptrace.go:74]: We were formerly using `r.URL.Path`. Why are we changing to this? Can we do it in a separate PR?

File: contrib/internal/httptrace/httptrace.go (hunk 21 above)
Comment (gbbr) [reply contrib/internal/httptrace/httptrace.go:27]: If it does, then yes, but not today.

File: contrib/internal/httptrace/httptrace.go (hunk 23 above)
Comment (Julio-Guerra) [reply contrib/internal/httptrace/httptrace.go]: Reverted in https://github.com/DataDog/dd-trace-go/pull/1286/commits/8c2a98a95de1fb526ad383b5fb3eb3f1a0eb0c79

**Review by gbbr — APPROVED** [review]
LGTM
